File: C_plus_plus/Valid_Palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        if(s.size()==0) return true;
        int i = 0, j = s.size()-1;
        while(i<=j){
            while(i<=j&&!isAlphanumeric(s[i])) ++i;
            while(i<=j&&!isAlphanumeric(s[j])) --j;
            if(i>j) return true;
            if(abs(s[i]-s[j])==0||abs(s[i]-s[j])==abs('a'-'A')){
                ++i;
                --j;
            }
            else return false;
        }
        return true;
    }
    bool isAlphanumeric(char c){
        if(c<='9'&&c>='0'||c<='z'&&c>='a'||c<='Z'&&c>='A') return true;
        return false;
    }
};

File: java/Valid_Palindrome.java
// Two pointer, one pass!
// Time is O(n), space is O(1).

public class Solution {
    public boolean isPalindrome(String s) {
        int i = 0, j = s.length() - 1;
        while (i < j) {
            char a = s.charAt(i), b = s.charAt(j);
            if (!isAlphaNum(a)) {
                ++i;
                continue;
            }
            if (!isAlphaNum(b)) {
                --j;
                continue;
            }
            if (a >= '0' && a <= '9') {
                if (a != b) return false;
            }
            else {
                if (Character.toLowerCase(a) != Character.toLowerCase(b))
                    return false;
            }
            ++i;
            --j;
        }
        return true;
    }
    public boolean isAlphaNum(char c) {
        char cc = Character.toLowerCase(c);
        if (c >= '0' && c <= '9' || cc >= 'a' && cc <= 'z')
            return true;
        return false;
    }
}
